src/cluster.c

```diff
@@ -1213,8 +1213,10 @@ void clusterRedirectClient(client *c, clusterNode *n, int hashslot, int error_co
  * returns 1. Otherwise 0 is returned and no operation is performed. */
 int clusterRedirectBlockedClientIfNeeded(client *c) {
     clusterNode *myself = getMyClusterNode();
-    if (c->flags & CLIENT_BLOCKED && (c->bstate.btype == BLOCKED_LIST || c->bstate.btype == BLOCKED_ZSET ||
-                                      c->bstate.btype == BLOCKED_STREAM || c->bstate.btype == BLOCKED_MODULE)) {
+    if (c->flags & CLIENT_BLOCKED &&
+        (c->bstate.btype == BLOCKED_LIST || c->bstate.btype == BLOCKED_ZSET || c->bstate.btype == BLOCKED_STREAM ||
+         c->bstate.btype == BLOCKED_MODULE) &&
+        !clientHasModuleAuthInProgress(c)) {
```

Comment:
Maybe we should update the top comment to reflect upon the additional condition.

Reply:
I think the top comment is fine as is. It's about clients being blocked on keys in a hash slot, i.e. on a blocking command that has keys. Auth isn't about keys.

Can we make the check more to the point here? We can check if the client is blocked on keys. AUTH doesn't have keys, so the dict is zero in this case. Additionally, this may cover other cases too, like blocked in module for other reasons. WDYT?

```suggestion
        dictSize(c->bstate.keys) != 0) {
```

Reply:
> We can check if the client is blocked on keys. AUTH doesn't have keys, so the dict is zero in this case. Additionally, this may cover other cases too, like blocked in module for other reasons. WDYT?

I think this may have been one of the intentions of this function.

A few lines below, we see this:
```
        /* If the cluster is down, unblock the client with the right error.
         * If the cluster is configured to allow reads on cluster down, we
         * still want to emit this error since a write will be required
         * to unblock them which may never come.  */
        if (!isClusterHealthy()) {
            clusterRedirectClient(c, NULL, 0, CLUSTER_REDIR_DOWN_STATE);
            return 1;
        }

        /* If the client is blocked on module, but not on a specific key,
         * don't unblock it (except for the CLUSTER_FAIL case above). */
        if (c->bstate.btype == BLOCKED_MODULE && !moduleClientIsBlockedOnKeys(c)) return 0;
```

Maybe we can just move this up (before the first check)